gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/configureJvmApplication.kt

```diff
@@ -430,27 +434,41 @@ private fun JvmApplicationContext.configureRunTask(
         add("-D$APP_RESOURCES_DIR=${appResourcesDir.absolutePath}")
     }
     exec.args = app.args
-    exec.useAppRuntimeFiles { (runtimeJars, _) ->
-        classpath = runtimeJars
+
+    if (runProguard != null) {
+        exec.dependsOn(runProguard)
+        exec.classpath = project.fileTree(runProguard.flatMap { it.destinationDir })
+    } else {
+        exec.useAppRuntimeFiles { (runtimeJars, _) ->
+            classpath = runtimeJars
+        }
     }
 }
 
-private fun JvmApplicationContext.configurePackageUberJarForCurrentOS(jar: Jar) {
+private fun JvmApplicationContext.configurePackageUberJarForCurrentOS(
+    jar: Jar,
+    runProguard: Provider<AbstractProguardTask>?
+) {
     fun flattenJars(files: FileCollection): FileCollection =
         jar.project.files({
             files.map { if (it.isZipOrJar()) jar.project.zipTree(it) else it }
         })
 
-
-    jar.useAppRuntimeFiles { (runtimeJars, _) ->
-        from(flattenJars(runtimeJars))
+    if (runProguard != null) {
+        jar.dependsOn(runProguard)
+        jar.from(flattenJars(project.fileTree(runProguard.flatMap { it.destinationDir })))
```

Comment:
After some tests i've found that `from()` method executes eagerly before the execution phase when there are no proguard processed files yet, so it remembers the empty files list.

I've tried to play with `fileTree()`, `zipTree()`, `files()` and closures for creating the lazy evaluated file tree of zip trees from files from `destinationDir`, but with no luck.

The only working solution that I found was this:
```kotlin
val fs: FileSystemOperations
val archive: ArchiveOperations

fun flattenJars(files: FileCollection) = files.map { if (it.isZipOrJar()) archive.zipTree(it) else it }

val proguardDestinationDir = jar.project.fileTree(runProguard.flatMap { it.destinationDir })
val uberjarFlattenedSourcesDir = appTmpDir.dir("uberjar")
jar.doFirst {
    fs.clearDirs(uberjarFlattenedSourcesDir)
    fs.copy {
        it.duplicatesStrategy = DuplicatesStrategy.EXCLUDE
        it.from(flattenJars(proguardDestinationDir))
        it.into(uberjarFlattenedSourcesDir)
    }
}
jar.from(uberjarFlattenedSourcesDir)
```
It works fine, but has some disadvantages:
- Uses temp directory for proguard processed jars unpacking and flattening instead of `zipTree`
- I don't know where to place and inject `FileSystemOperations` and `ArchiveOperations` better.
Usage of corresponding `Project` methods is forbidden because of gradle configuration cache.

Maybe you can suggest something?

Reply:
yes, you are right, but I'd recommend to register a custom task to flat an uberjar before the jar task. And there you may `inject` needed operations.

Reply:
Fixed by a separate task.

Reply:
After testing in prod i've found one more problem - on Windows unpacking to the file system removes some files because of case insensitive FS.
As a workaround we can use `-dontusemixedcaseclassnames` in proguard config, but adding it to the plugin defaults looks like a bad idea.
How else we can fix it?
Maybe completely replace the `Jar` task with smthng more powerful like a ShadowJar plugin?